modules/optimization/src/vectorpostprocessors/OptimizationFunctionInnerProductHelper.C

```diff
@@ -0,0 +1,115 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "OptimizationFunctionInnerProductHelper.h"
+
+// MOOSE includes
+#include "InputParameters.h"
+#include "MooseError.h"
+#include "FEProblemBase.h"
+
+// Optimization includes
+#include "OptimizationFunction.h"
+
+InputParameters
+OptimizationFunctionInnerProductHelper::validParams()
+{
+  InputParameters params = emptyInputParameters();
+  params.addRequiredParam<FunctionName>("function", "Optimization function.");
+  params.addParam<Real>(
+      "reverse_time_end",
+      0.0,
+      "End time used for reversing the time integration when evaluating function derivative.");
+  return params;
+}
+
+OptimizationFunctionInnerProductHelper::OptimizationFunctionInnerProductHelper(
+    const InputParameters & parameters)
+  : _ip_problem(*parameters.getCheckedPointerParam<FEProblemBase *>("_fe_problem_base")),
+    _function(dynamic_cast<const OptimizationFunction *>(&_ip_problem.getFunction(
+        parameters.get<FunctionName>("function"), parameters.get<THREAD_ID>("_tid")))),
+    _reverse_time_end(parameters.get<Real>("reverse_time_end")),
+    _simulation_time(_ip_problem.getMooseApp().getStartTime())
+{
+  if (!_function)
+    mooseError("Function requested by ",
+               parameters.get<std::string>("_type"),
+               " must be an OptimizationFunction.");
+}
+
+void
+OptimizationFunctionInnerProductHelper::setCurrentTime(Real time, Real dt)
+{
+  // We are saving the integral for each time step.
+  // So create or grab the vector if the time has changed or this is our first time here.
+  _curr_time_ip = nullptr;
+  for (auto & pr : _time_ip)
+    if (MooseUtils::relativeFuzzyEqual(time, pr.first))
+      _curr_time_ip = &pr.second;
+  if (!_curr_time_ip)
+  {
+    _time_ip.emplace_back(time, std::vector<Real>());
+    _curr_time_ip = &_time_ip.back().second;
+  }
+  _curr_time_ip->clear();
+
+  _actual_time =
+      MooseUtils::absoluteFuzzyEqual(_reverse_time_end, 0.0) ? time : _reverse_time_end - time + dt;
+}
+
+void
+OptimizationFunctionInnerProductHelper::update(const Point & q_point, Real q_inner_product)
```

Comment:
I m not sure about this update name for the routine. Seems to just be a part of the inner product?

Reply:
I think it is an apt name since it is updating the members of this class with `adjoint[qp] * forward[qp]`. Do you have a better suggestion?

Reply:
No